src/libstd/macros.rs

```diff
@@ -213,6 +213,27 @@ macro_rules! eprintln {
     ($fmt:expr, $($arg:tt)*) => (eprint!(concat!($fmt, "\n"), $($arg)*));
 }
 
+#[macro_export]
+#[unstable(feature = "await_macro", issue = "50547")]
+#[allow_internal_unstable]
+#[allow_internal_unsafe]
+macro_rules! await {
+    ($e:expr) => { {
+        let mut pinned = $e;
+        let mut pinned = unsafe { $crate::mem::PinMut::new_unchecked(&mut pinned) };
+        loop {
+            match $crate::future::get_task_cx(|cx|
+                $crate::future::Future::poll(pinned.reborrow(), cx))
```

Comment:
Is there any risk here wrt `reborrow` methods? Should this be calling `$crate::mem::PinMut::reborrow` explicitly?

Also, `get_task_cx` doesn't appear to need its full generality, there could instead be a function that takes the `PinMut` and calls `Future::poll` itself. Has the slight advantage of not generating a closure per use of `await`.